File: python/sort/timsort.py
def short_tim_sort(arr):
    return sorted(arr)

def tim_sort(arr):
    if not arr or len(arr) <= 1:
        return arr
    
    n = len(arr)
    
    # Tamaño de ejecución mínimo
    min_run = 32
    
    # Paso 1: Ordenar subarrays pequeños con insertion sort
    for start in range(0, n, min_run):
        end = min(start + min_run - 1, n - 1)
        insertion_sort(arr, start, end)
    
    # Paso 2: Combinar los subarrays ordenados
    size = min_run
    while size < n:
        for left in range(0, n, 2 * size):
            mid = min(n - 1, left + size - 1)
            right = min(n - 1, left + 2 * size - 1)
            
            if mid < right:
                merge(arr, left, mid, right)
        
        size *= 2
    
    return arr

def insertion_sort(arr, left, right):
    """Ordenamiento por inserción para un rango específico."""
    for i in range(left + 1, right + 1):
        key = arr[i]
        j = i - 1
        while j >= left and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

def merge(arr, left, mid, right):
    """Combina dos subarrays ordenados."""
    len1, len2 = mid - left + 1, right - mid
    left_arr = arr[left:mid + 1]
    right_arr = arr[mid + 1:right + 1]
    
    i = j = 0
    k = left
    
    while i < len1 and j < len2:
        if left_arr[i] <= right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
    
    while i < len1:
        arr[k] = left_arr[i]
        i += 1
        k += 1
    
    while j < len2:
        arr[k] = right_arr[j]
        j += 1
        k += 1

if __name__ == "__main__":
    array = [5, 2, 9, 1, 5, 6]
    
    # Implementacion automatica
    sorted_array1 = short_tim_sort(array.copy())
    print("Implementacion automatica:", sorted_array1)
    
    # Implementacion manual
    sorted_array2 = tim_sort(array.copy())
    print("Implementacion manual:", sorted_array2)

File: java/sort/TimSort.java
package sort;

import java.util.Arrays;

public class TimSort {
    public static void main(String[] args) {
        int[] array = {5, 2, 9, 1, 5, 6};
        
        // Implementacion automatica
        int[] sortedArray1 = shortTimSort(array.clone());
        System.out.println("Implementacion automatica: " + Arrays.toString(sortedArray1));
        
        // Implementacion manual
        int[] sortedArray2 = timSort(array.clone());
        System.out.println("Implementacion manual: " + Arrays.toString(sortedArray2));
    }
    
    public static int[] shortTimSort(int[] array) {
        Arrays.sort(array);
        return array;
    }

    public static int[] timSort(int[] arr) {
        if (arr == null || arr.length <= 1) return arr;
        
        int n = arr.length;
        
        // Tamaño de ejecución mínimo
        int minRun = 32;

        for (int i = 0; i < n; i += minRun) {
            insertionSort(arr, i, Math.min(i + minRun - 1, n - 1));
        }
        
        // Combinar con merge
        for (int size = minRun; size < n; size = 2 * size) {
            for (int left = 0; left < n; left += 2 * size) {
                int mid = left + size - 1;
                int right = Math.min(left + 2 * size - 1, n - 1);
                
                if (mid < right) {
                    merge(arr, left, mid, right);
                }
            }
        }
        
        return arr;
    }
    
    private static void insertionSort(int[] arr, int left, int right) {
        for (int i = left + 1; i <= right; i++) {
            int key = arr[i];
            int j = i - 1;
            
            while (j >= left && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }
    
    private static void merge(int[] arr, int left, int mid, int right) {
        int[] temp = new int[right - left + 1];
        int i = left, j = mid + 1, k = 0;
        
        while (i <= mid && j <= right) {
            if (arr[i] <= arr[j]) {
                temp[k++] = arr[i++];
            } else {
                temp[k++] = arr[j++];
            }
        }
        
        while (i <= mid) {
            temp[k++] = arr[i++];
        }
        
        while (j <= right) {
            temp[k++] = arr[j++];
        }
        
        for (i = 0; i < k; i++) {
            arr[left + i] = temp[i];
        }
    }
}
